fairscale/nn/pipe/phony.py

```diff
@@ -55,7 +55,9 @@ def forward(ctx, input):
         phony = _phonies[key]
     except KeyError:
         with use_stream(default_stream(device)):
-            phony = torch.empty(0, device=device, requires_grad=requires_grad)
+            # Creating phony with size 1 instead of zero, since currently
+            # tensorpipe does not work with tensors of size zero.
+            phony = torch.empty(1, device=device, requires_grad=requires_grad)
```

Comment:
Why do empty tensors need to be sent?

Reply:
These phony's are used to inject dependencies in auto-grad engine between the tensors for batch i in partition j and and tensors for batch i+1 in partition j-1. So they need to be sent over to the remote partition j-1. This is to make sure in backward path, first  partition j-1 finishes processing batch i+1, and then  grads for the phony are transmitted back, and then the grads for batch i in partition j is sent over to partition j-1 (to prevent queuing in partition j-1).

Reply:
It should be a no-op change for single process pipe. The tensor is never moved between devices, and no real operation is run on them.